qubes-trust-daemon.cpp

```diff
@@ -0,0 +1,355 @@
+                        untrusted_buffer.insert(fullpath);
+                    }
+                }
+
+                if (event->mask & IN_MOVED_TO) {
+                    if (event->mask & IN_ISDIR) {
+                        printf("%d DIR::%s CREATED\n", event->wd, fullpath.c_str());
+                        place_watch_on_dir_and_subdirs(fullpath.c_str());
+                    } else {
+                        printf("%d FILE::%s CREATED\n", event->wd, fullpath.c_str());
+                        // Mark file to be set as untrusted
+                        untrusted_buffer.insert(fullpath);
+                    }
+                }
+            }
+
+            if (event->mask & IN_MOVED_FROM) {
+                if (event->mask & IN_ISDIR) {
+                    printf("%d REMOVE DIR::%s\n", event->wd, fullpath.c_str());
+
+                    // TODO: Recursive rm watch
+                    inotify_rm_watch(watch_fd, event->wd);
+                } else {
+                    printf("%d FILE::%s MOVED OUT\n", event->wd, fullpath.c_str());
+                }
+            }
+
+            if (event->mask & IN_MODIFY) {
+                if (event->mask & IN_ISDIR) {
+                    printf("%d DIR::%s MODIFIED\n", event->wd, fullpath.c_str());
+                } else {
+                    printf("%d FILE::%s MODIFIED\n", event->wd, fullpath.c_str());
+                }
+            }
+
+            i += EVENT_SIZE + event->len;
+        }
+    }
+}
+
+/*
+ * Get a list of untrusted directories to watch
+ * from qvm-file-trust's output
+ */
+std::set<std::string> get_untrusted_dir_list() {
+    std::set<std::string> rules;
+
+    FILE* fp = popen("/usr/bin/qvm-file-trust -p", "r");
+    char buf[1024*1024];
+
+    fread(buf, 1, sizeof(buf), fp);
+    std::string rules_str = buf;
+
+    std::stringstream ss(rules_str);
+    std::string to;
+
+    while (std::getline(ss, to, '\n')) {
+        rules.insert(to);
+    }
+    return rules;
+}
+
+/*
+ * Run every few seconds and send untrusted_buffer to handler method 
+ */
+void *set_trust_on_timer(void*) {
+    while(1) {
+        sleep(UNTR_MARK_PERIOD);
+        mark_files_as_untrusted(untrusted_buffer);
+    }
+    return 0;
+}
+
+int main(void) {
+    // TODO: Set up logging - log to syslog
+
+    // Initialize inotify
+    watch_fd = inotify_init();
+    if (watch_fd < 0) {
+        std::cerr << "Unable to initialize inotify" << std::endl;
+    }
+
+    // Get a list of all untrusted directories
+    std::set<std::string> untrusted_dirs = get_untrusted_dir_list();
+
+    // Add a watch to each untrusted directory and their subdirectories
+    std::set<std::string>::iterator it;
+    std::string dir;
+    for (it = untrusted_dirs.begin(); it != untrusted_dirs.end(); ++it) {
+        dir = *it;
+        place_watch_on_dir_and_subdirs(dir.c_str());
+    }
+
+    // Start untrusted_buffer monitor
+    pthread_t tid;
+    pthread_create(&tid, NULL, &set_trust_on_timer, NULL);
```

Comment:
Ough. This looks like asking for troubles. First, are you sure that std::set is thread-safe? Then, there are probably a lot of race conditions, at least `untrusted_buffer.clear()` in `mark_files_as_untrusted()` can be called after adding some entries by the other thread.

Better call `mark_files_as_untrusted()` from the main thread from time to time - for example after reading every batch of events.

Reply:
I wonder if maybe this shouldn't be poll(2)/select(2)/epoll(2)/kqueue(2) based instead of multi-threaded at all to begin with.

Reply:
Yes, I was thinking about dropping this additional thread. And since there is only one FD to look for (inotify), simple read(2) loop should be enough.

Reply:
Yeah, the reason for waiting in 1 second intervals is to let the events build up before spinning up the python interpreter and processing them, but I agree that a separate thread is probably overkill for this.

@marmarek Would a read loop allow events to build up or just act on them immediately?

@jpouellet That might be a better solution if we can't do it in an event-based manner (and instead do it on a timer like we currently are).

Reply:
Depending on the speed at which they are produced. read() will return multiple events, if there are available. I guess, in the worst case, they will pile up during first invocation of python script.

Reply:
Changed in [2d24c06](https://github.com/anoadragon453/qubes-file-trust/commit/2d24c06ed1c42bdc13fa11bc262122e92ece4585). Instead of batching calls based on a timer, we call the python cli, wait for it to exit, and only then check if there's another batch to process.

Any new inotify events also trigger this process.